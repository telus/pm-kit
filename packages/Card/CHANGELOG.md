# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [2.0.0](https://github.com/telus/pm-kit/compare/@pm-kit/card@1.1.1...@pm-kit/card@2.0.0) (2020-02-11)


### Features

* **card:** updated details bar to be customizable ([#54](https://github.com/telus/pm-kit/issues/54)) ([f5deab8](https://github.com/telus/pm-kit/commit/f5deab8ba68c069e8aea8d81a399ca976af48f58))


### BREAKING CHANGES

* **card:** changed `isExpandable` prop to `expandable`, which is now an object rather than a boolean





## [1.1.1](https://github.com/telus/pm-kit/compare/@pm-kit/card@1.1.0...@pm-kit/card@1.1.1) (2020-02-10)


### Bug Fixes

* **card:** update card details-bar styling ([#53](https://github.com/telus/pm-kit/issues/53)) ([802bf77](https://github.com/telus/pm-kit/commit/802bf77bd203d6381b9b513d1ef2e0d4bab0674e))





# [1.1.0](https://github.com/telus/pm-kit/compare/@pm-kit/card@1.0.0...@pm-kit/card@1.1.0) (2020-01-27)


### Features

* **card:** updated card styling and documentation ([#39](https://github.com/telus/pm-kit/issues/39)) ([53a4dba](https://github.com/telus/pm-kit/commit/53a4dba292c107d3da9218a83dd8086ae01ecb5f))





# 1.0.0 (2020-01-24)

**Note:** Version bump only for package @pm-kit/card
