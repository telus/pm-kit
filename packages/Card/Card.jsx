import React, { useState, useEffect, useRef } from 'react'
import PropTypes from 'prop-types'
import { css } from '@emotion/core'
import { motion, AnimatePresence } from 'framer-motion'
import { offWhite, parkGreen, lilyGreen } from '@pm-kit/colours'
import { weight } from '@pm-kit/typography'
import Paragraph from '@pm-kit/paragraph'

import downArrow from '../../shared/svg/arrow-show.svg'
import upArrow from '../../shared/svg/arrow-hide.svg'
import checkmark from '../../shared/png/Checkmark/verified@3x.png'
import oval from '../../shared/png/EmptyOval/oval@3x.png'

const card = css`
  width: 100%;
  display: flex;
  flex-direction: column;
  background-color: ${offWhite};
  border-width: 3px;
  border-radius: 12px;
  border-style: solid;
  overflow: hidden;
  position: relative;
  border-color: ${parkGreen};
  flex-basis: '40%';
  flex-grow: '1';
  margin-bottom: '32px';
  &:hover {
    box-shadow: 0 2px 18px 0 rgba(0, 0, 0, 0.25);
  }
`

const selectedCard = css`
  background-color: ${lilyGreen};
  box-shadow: 0 2px 18px 0 rgba(0, 0, 0, 0.25);
`

const titleStyle = css`
  color: ${parkGreen};
  font-weight: ${weight.bold};
  padding-left: 8px;
  width: 100%;
`

const paddingLeft = css`
  padding-left: 8px;
`

const line = css`
  border-style: solid;
  border-color: ${parkGreen};
  border-bottom-width: 1px;
  margin-left: 35px;
  margin-right: 35px;
  &:focus {
    outline: none;
  }
`

const detailsBar = css`
  display: flex;
  flex-direction: row;
  font-weight: ${weight.bold};
  justify-content: space-between;
  text-decoration: underline;
  padding: 16px 23px 23px 23px;
  cursor: pointer;
  background-color: inherit;
  border: none;
`

const cardRowContainer = css`
  display: flex;
  flex-direction: row;
  align-items: flex-end;
  margin-bottom: 24px;
  margin-top: 8px;
`

const selectOptionContainer = css`
  padding: 23px 23px 0px 23px;
  cursor: pointer;
  display: flex;
  flex-direction: column-reverse;
`

const selectOptionContainerNoCurser = css`
  padding: 23px 23px 0px 23px;
  cursor: default;
`

const cardDetails = css`
  color: ${parkGreen};
  padding: 0px 23px 0px 23px;
  overflow: hidden;
`

const detailSection = css`
  padding-top: 16px;
  padding-bottom: 24px;
  display: flex;
  flex-direction: column;
  & > ul {
    list-style: disc;
    list-style-position: inside;
  }
  & > ul > li > span {
    margin-left: -12px;
  }
`

const selectOption = css`
  display: flex;
  flex-direction: row-reverse;
  font-weight: ${weight.bold};
`

const planSelected = css`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  background-color: inherit;
  border: none;
`

const cardSelectedText = css`
  padding-right: 5px;
`
const cardNotSelectedText = css`
  text-decoration: underline;
  padding-right: 8px;
`
const cardContainerDetails = css`
  display: flex;
  justify-content: space-between;
  flex-direction: column;
`

const planVariant = {
  hidden: { opacity: 0, y: -10 },
  visible: {
    opacity: 1,
    y: 0,
    transition: { y: { type: 'spring', damping: 100, velocity: 500 }, opacity: { duration: 0.8 } },
  },
}

const Card = ({ expandable, placeholder, title, subtitle, children, onClick, isSelected, styles, selectable }) => {
  const [openCard, setOpenCard] = useState(false)

  const detailsRef = useRef()
  const cardStyles = [card]

  const detailsBarStyles = [detailsBar]
  const selectOptionContainerStyles = [selectOptionContainer]

  if (!selectable) {
    selectOptionContainerStyles.push(selectOptionContainerNoCurser)
  }

  if (isSelected) {
    cardStyles.push(selectedCard)
  }

  const onCardClick = () => {
    onClick()
  }

  const toggleOpenCard = () => {
    setOpenCard(!openCard)
    if (openCard) {
      detailsRef.current.blur()
    } else if (detailsRef && detailsRef.current) {
      detailsRef.current.focus()
    }
  }

  useEffect(() => {
    if (!expandable) {
      setOpenCard(true)
    }
  }, [expandable])

  let arrowImage = downArrow

  if (openCard) {
    arrowImage = upArrow
  } else {
    arrowImage = downArrow
  }

  if (styles) {
    if (styles.detailsBarStylingOverride) {
      detailsBarStyles.push(styles.detailsBarStylingOverride)
    }
    if (styles.selectOptionContainerStylingOverride) {
      selectOptionContainerStyles.push(styles.selectOptionContainerStylingOverride)
    }
  }

  const paragraphSmallSizeStyle = styles && styles.fontSizeSmallOverride ? styles.fontSizeSmallOverride : ''
  const paragraphLargeSizeStyle = styles && styles.fontSizeLargeOverride ? styles.fontSizeLargeOverride : ''
  const paragraphInheritSizeStyle = styles && styles.fontSizeSmallOverride ? styles.fontSizeSmallOverride : 'inherit'

  return (
    <motion.div css={cardStyles} variants={planVariant}>
      <div css={cardContainerDetails}>
        <div css={selectOptionContainerStyles} onClick={selectable ? onCardClick : undefined}>
          <div css={cardRowContainer} tabIndex="0">
            <div>{placeholder && placeholder}</div>
            <Paragraph css={titleStyle} size={paragraphLargeSizeStyle}>
              {title}
            </Paragraph>
            <p css={paddingLeft}>{subtitle}</p>
          </div>
          {selectable && (
            <div css={selectOption}>
              {isSelected && (
                <button name="plan" type="button" css={planSelected}>
                  <Paragraph weight={weight.bold} css={cardSelectedText} size={paragraphSmallSizeStyle}>
                    {selectable.selectedText}
                  </Paragraph>
                  <img src={checkmark} height="18px" width="18px" alt="selected" />
                </button>
              )}
              {!isSelected && (
                <button name="plan" type="button" css={planSelected}>
                  <Paragraph weight={weight.bold} css={cardNotSelectedText} size={paragraphSmallSizeStyle}>
                    {selectable.unSelectedText}
                  </Paragraph>
                  <img src={oval} height="18px" width="18px" alt="unselected" />
                </button>
              )}
            </div>
          )}
        </div>
        {children && (
          <>
            <div css={line} ref={detailsRef} tabIndex="-1" />
            <AnimatePresence>
              {openCard && (
                <motion.div
                  css={cardDetails}
                  initial={{ height: 0 }}
                  animate={{ height: 'auto' }}
                  exit={{ height: '0px' }}
                >
                  <div css={detailSection}>{children}</div>
                </motion.div>
              )}
            </AnimatePresence>
          </>
        )}
        {expandable && (
          <button name="details" type="button" css={detailsBarStyles} onClick={toggleOpenCard}>
            <Paragraph weight={weight.bold} size={paragraphInheritSizeStyle}>
              {openCard && <>{expandable.collapse}</>}
              {!openCard && <>{expandable.details}</>}
            </Paragraph>
            <img src={arrowImage} alt="arrow" />
          </button>
        )}
      </div>
    </motion.div>
  )
}

Card.propTypes = {
  /**
   * The title of the Card.
   */
  title: PropTypes.string,
  /**
   * The subtitle of the Card.
   */
  subtitle: PropTypes.string,
  /**
   * Any HTML element that you want to be rendered
   */
  placeholder: PropTypes.node,
  /**
   * The Details of the Card. Having children of any type renders an expandable details row.
   * With the contents of the child supplied to it.
   */
  children: PropTypes.node,
  /**
   * This is a fuction that is passed into the component that executes when the component is clicked.
   */
  onClick: PropTypes.func,
  /**
   * A boolean that if true renders the selected state of the card. If false renders the default state.
   */
  isSelected: PropTypes.bool,
  /**
   * Contains an object with the selected and unselected text in the top right of the Card.
   * This allows the card to be selectable.
   */
  selectable: PropTypes.PropTypes.exact({
    selectedText: PropTypes.string.isRequired,
    unSelectedText: PropTypes.string.isRequired,
  }),
  /**
   * Contains an object with the details and collapse text to be shown in the the details bar of the Card.
   * This allows the card to be expandable.
   */
  expandable: PropTypes.exact({
    details: PropTypes.string.isRequired,
    collapse: PropTypes.string.isRequired,
  }),
  /**
   * Customizes the Card according to your needs.
   * Accepts an object of styles in the structure below.
   * {
   *  selectOptionContainerStylingOverride: {},
   *  detailsBarStylingOverride: {},
   *  fontSizeSmallOverride: "",
   *  fontSizeLargeOverride: "",
   * }
   * */
  styles: PropTypes.object,
}

Card.defaultProps = {
  title: undefined,
  subtitle: undefined,
  placeholder: undefined,
  children: undefined,
  onClick: undefined,
  isSelected: undefined,
  selectable: undefined,
  expandable: undefined,
}

export default Card
