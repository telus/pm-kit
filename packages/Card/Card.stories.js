import React, { useState } from 'react'
import { withKnobs, text, select, object } from '@storybook/addon-knobs'
import Card from './Card'
import { version } from './package.json'

export default {
  title: 'Design System|Card',
  component: Card,
  parameters: {
    componentSubtitle: `version ${version}`,
  },
  decorators: [withKnobs],
}

const placeholderHTML = <p style={{ border: '1px solid', height: '50px', width: '50px' }}></p>

const detailsHtml = <p> This Card can Expand but cannot be Selected</p>

export const Primary = () => {
  return <Card title="Default Title" subtitle="Default Subtitle" />
}

export const Selectable = () => {
  const [selectCard, setSelectedCard] = useState(false)
  const id = 1
  return (
    <Card
      title="Default Title"
      subtitle="Default Subtitle"
      selectable={{ selectedText: 'Selected', unSelectedText: 'Select' }}
      onClick={() => setSelectedCard(id)}
      isSelected={selectCard === id}
    />
  )
}

export const Expandable = () => {
  return (
    <Card title="Default Title" subtitle="Default Subtitle" expandable={{ details: 'Details', collapse: 'Collapse' }}>
      {detailsHtml}
    </Card>
  )
}

export const NotExpandable = () => {
  return (
    <Card title="Default Title" subtitle="Default Subtitle" expandable={{ details: 'Details', collapse: 'Collapse' }}>
      <p> This Card cannot Expand and cannot be Selected</p>
    </Card>
  )
}

export const WithPlaceholder = () => {
  const [selectCard, setSelectedCard] = useState()
  const id = 1
  return (
    <Card
      placeholder={placeholderHTML}
      title="Default Title"
      subtitle="Default Subtitle"
      selectable={{ selectedText: 'Selected', unSelectedText: 'Select' }}
      isSelected={selectCard === id}
      onClick={() => setSelectedCard(id)}
      expandable={{ details: 'Details', collapse: 'Collapse' }}
    >
      <p>Default Details</p>
    </Card>
  )
}

export const MultipleCards = () => {
  const [selectCard, setSelectedCard] = useState()
  const idCard1 = 1
  const idCard2 = 2
  return (
    <div style={{ display: 'flex', flexDirection: 'row', justifyContent: 'space-between' }}>
      <div style={{ width: '40%' }}>
        <Card
          title="Card 1"
          selectable={{ selectedText: 'Selected', unSelectedText: 'Select' }}
          isSelected={selectCard === idCard1}
          onClick={() => setSelectedCard(idCard1)}
          expandable={{ details: 'Details', collapse: 'Collapse' }}
        >
          <>
            <p>Details for card1</p>
            <ul>
              <li>Detail 1</li>
              <li>Detail 2</li>
              <li>Detail 3</li>
              <li>Detail 4</li>
              <li>Detail 5</li>
            </ul>
          </>
        </Card>
      </div>
      <div style={{ width: '40%' }}>
        <Card
          title="Card 2"
          selectable={{ selectedText: 'Selected', unSelectedText: 'Select' }}
          isSelected={selectCard === idCard2}
          onClick={() => setSelectedCard(idCard2)}
          expandable={{ details: 'Details', collapse: 'Collapse' }}
        >
          <>
            <p>Details for card2</p>
            <ul>
              <li>Detail 1</li>
              <li>Detail 2</li>
              <li>Detail 3</li>
              <li>Detail 4</li>
              <li>Detail 5</li>
            </ul>
          </>
        </Card>
      </div>
    </div>
  )
}

export const Playground = () => {
  const [selectCard, setSelectedCard] = useState()
  const id = 1
  return (
    <Card
      title={text('Title', 'Title')}
      subtitle={text('Subtitle', 'Subtitle')}
      placeholder={select('Placeholder', [undefined, 'Placeholder'])}
      onClick={() => setSelectedCard(id)}
      isSelected={selectCard === id}
      selectable={object('Selected Text', { unSelectedText: 'Select', selectedText: 'Selected' })}
      expandable={object('Expandable Text', { details: 'Details', collapse: 'Collapse' })}
    >
      {text('Edit Details', 'Edit Details')}
    </Card>
  )
}

Playground.story = {
  name: 'playground',
  parameters: { docs: { page: null, disable: true } },
}
