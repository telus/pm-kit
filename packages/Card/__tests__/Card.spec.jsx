import React from 'react'
import { render } from 'enzyme'

import Card from '../Card'

describe('Card', () => {
  it('renders', () => {
    const card = render(<Card title="title" subtitle="subtitle" />)

    expect(card).toMatchSnapshot()
  })
})
