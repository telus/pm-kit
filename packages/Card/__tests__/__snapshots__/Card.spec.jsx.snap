// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders 1`] = `
.emotion-5 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background-color: #FBF5EB;
  border-width: 3px;
  border-radius: 12px;
  border-style: solid;
  overflow: hidden;
  position: relative;
  border-color: #034045;
  -webkit-flex-basis: '40%';
  -ms-flex-preferred-size: '40%';
  flex-basis: '40%';
  -webkit-box-flex: '1';
  -webkit-flex-grow: '1';
  -ms-flex-positive: '1';
  flex-grow: '1';
  margin-bottom: '32px';
}

.emotion-5:hover {
  box-shadow: 0 2px 18px 0 rgba(0,0,0,0.25);
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-3 {
  padding: 23px 23px 0px 23px;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column-reverse;
  -ms-flex-direction: column-reverse;
  flex-direction: column-reverse;
  padding: 23px 23px 0px 23px;
  cursor: default;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  margin-bottom: 24px;
  margin-top: 8px;
}

.emotion-0 {
  text-align: left;
  width: 100%;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 300;
  color: #034045;
  color: #034045;
  font-weight: 500;
  padding-left: 8px;
  width: 100%;
}

.emotion-1 {
  padding-left: 8px;
}

<div
  class="emotion-5"
>
  <div
    class="emotion-4"
  >
    <div
      class="emotion-3"
    >
      <div
        class="emotion-2"
        tabindex="0"
      >
        <div />
        <p
          class="emotion-0"
        >
          title
        </p>
        <p
          class="emotion-1"
        >
          subtitle
        </p>
      </div>
    </div>
  </div>
</div>
`;
